lib/booking_manager.rb

```
@@ -0,0 +1,184 @@
+        raise ArgumentError.new "Invalid date range. Start date must be before end date, both in format of 'Month dd, yyyy'. "
+      end
+    end
+
+
+    # Create array of all dates from start date to end date
+    def determine_date_range(start_date, end_date)
+      check_dates(start_date, end_date)
+
+      res_start_date = Date.parse(start_date)
+      res_end_date = Date.parse(end_date)
+
+      date_range = []
+      search_date = res_start_date
+
+      until search_date > res_end_date
+        date_range << search_date
+        search_date += 1
+      end
+      return date_range
+    end
+
+    # Method to list all reservation instances
+    def list_reservations
+      return @reservations
+    end
+
+
+    # Method to list all rooms in hotel
+    def list_rooms
+      return @rooms
+    end
+
+
+    # Method to get total cost of reservation
+    def get_reservation_cost(nights, cost_per_night)
+      total_cost = nights * cost_per_night
+      return total_cost
+    end
+
+    # Return array reservations with matching date from room calendar hash
+    def find_reservations_on_date(date, calendar)
+      search_date = Date.parse(date)
+      found_reservations = []
+
+       calendar.each do |room, info|
+         info.each do |date, reservation|
+           if date == search_date
+             found_reservations << reservation
+           else
+            next
+          end
+        end
+       end
+       # Add return message for no reservations found?
+       return found_reservations
+    end
+
+    def find_vacancies_on_date(date, calendar)
+      search_date = Date.parse(date)
+      found_vacancies = []
+
+      calendar.each do |room, info|
+        if info == nil
+          found_vacancies << room
+        else
+          info.any? {|date, reservation| date == search_date}? next : found_vacancies << room
+        end
+      end
+
+      return found_vacancies.empty? ? no_vacancies_message: found_vacancies
+    end
```

In general, it's considered bad practice to return a string in a failure case like this. One of the main reasons is that whoever calls this code will have to do a lot of work to determine whether or not it failed.

Instead you should either raise an exception, or (since this is supposed to return a collection) return an empty array. A simple implementation of the second option would be replacing line 143 with

```ruby
return found_vacancies
```